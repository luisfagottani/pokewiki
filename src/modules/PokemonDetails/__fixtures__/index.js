const details = {
  name: 'bulbasaur',
  abilities: [
    {
      ability: {
        name: 'overgrow',
        url: 'https://pokeapi.co/api/v2/ability/65/',
      },
      is_hidden: false,
      slot: 1,
    },
    {
      ability: {
        name: 'chlorophyll',
        url: 'https://pokeapi.co/api/v2/ability/34/',
      },
      is_hidden: true,
      slot: 3,
    },
  ],
  height: 7,
  stats: [
    {
      base_stat: 45,
      effort: 0,
      stat: {
        name: 'hp',
        url: 'https://pokeapi.co/api/v2/stat/1/',
      },
    },
    {
      base_stat: 49,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
    {
      base_stat: 49,
      effort: 0,
      stat: {
        name: 'defense',
        url: 'https://pokeapi.co/api/v2/stat/3/',
      },
    },
    {
      base_stat: 65,
      effort: 1,
      stat: {
        name: 'special-attack',
        url: 'https://pokeapi.co/api/v2/stat/4/',
      },
    },
    {
      base_stat: 65,
      effort: 0,
      stat: {
        name: 'special-defense',
        url: 'https://pokeapi.co/api/v2/stat/5/',
      },
    },
    {
      base_stat: 45,
      effort: 0,
      stat: {
        name: 'speed',
        url: 'https://pokeapi.co/api/v2/stat/6/',
      },
    },
  ],
  weight: 69,
  id: 1,
  base_experience: 64,
};

export { details };
