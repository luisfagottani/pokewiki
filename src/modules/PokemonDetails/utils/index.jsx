const progressBarColor = (progress) => {
  if (progress < 60) {
    return 'blue';
  } else if (progress >= 60 && progress <= 70) {
    return 'yellow';
  } else {
    return 'red';
  }
};

export { progressBarColor };
