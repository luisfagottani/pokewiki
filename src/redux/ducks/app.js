import produce from 'immer';

export const REDUCER_NAME = 'app';

// TYPES
export const Types = {
  SET_GLOBAL_LOADING: `${REDUCER_NAME}/SET_GLOBAL_LOADING`,
};

// INITIAL STATE
const initialState = {
  isLoading: true,
  locale: 'pt-BR',
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.SET_GLOBAL_LOADING:
      draft.isLoading = action.payload;
      return draft;

    case Types.SET_USER_PROFILE:
      draft.profile = action.profile;
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS

export function defineGlobalLoading(status) {
  return {
    type: Types.SET_GLOBAL_LOADING,
    payload: status,
  };
}

// SELECTOR FUNCTIONS
export function getGlobalLoading(state) {
  return state[REDUCER_NAME].isLoading;
}

export function getLocale(state) {
  return state[REDUCER_NAME].locale;
}
