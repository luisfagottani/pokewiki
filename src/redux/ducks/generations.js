import produce from 'immer';

export const REDUCER_NAME = 'family';

// TYPES
export const Types = {
  FETCH_GENERATIONS: `${REDUCER_NAME}/FETCH_POKEMONS`,
  FETCH_GENERATIONS_BY_ID: `${REDUCER_NAME}/FETCH_GENERATIONS_BY_ID`,
  PERSIST_GENERATIONS: `${REDUCER_NAME}/PERSIST_GENERATIONS`,
};

// INITIAL STATE
const initialState = {
  generations: [],
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.PERSIST_GENERATIONS:
      draft.generations = { ...action.payload };
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS
export function fetchGenerations() {
  return {
    type: Types.FETCH_GENERATIONS,
  };
}

export function fetchGenerationsById(id) {
  return {
    type: Types.FETCH_GENERATIONS_BY_ID,
    payload: id,
  };
}

export function persistGenerations(pokemons) {
  return {
    type: Types.PERSIST_GENERATIONS,
    payload: pokemons,
  };
}

// SELECTORS
export function getGenerations(state) {
  return state[REDUCER_NAME]?.generations;
}
