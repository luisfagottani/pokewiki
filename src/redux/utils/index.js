const reduceSpecies = (results) =>
  results.reduce((current, next) => {
    const name = next?.name;
    if (!current[name]) {
      current[name] = { name };
    }
    return current;
  }, {});

export { reduceSpecies };
