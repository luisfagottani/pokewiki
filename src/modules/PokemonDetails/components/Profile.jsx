import { useIntl } from 'react-intl';

import TermsList from 'modules/shared/TermsList';
import { pokemonImage } from 'utils/functions';

import style from './Profile.module.scss';

const Profile = ({ name = '', details = undefined }) => {
  const intl = useIntl();

  return (
    <div className={style.profile} data-testid="profile">
      <figure className={style.profile__figure}>
        <figcaption className={style.profile__caption}>{name}</figcaption>
        {pokemonImage(name, style.profile__image)}
      </figure>

      <TermsList isLoading={!details}>
        <TermsList.List>
          <TermsList.Item term={intl.formatMessage({ id: 'details.profile.national_number' })} value={details?.id} />
          <TermsList.Item
            term={intl.formatMessage({ id: 'details.profile.base_experiene' })}
            value={details?.base_experience}
          />
          <TermsList.Item term={intl.formatMessage({ id: 'details.profile.height' })} value={details?.height} />
          <TermsList.Item term={intl.formatMessage({ id: 'details.profile.weight' })} value={details?.weight} />
        </TermsList.List>
      </TermsList>
    </div>
  );
};

export default Profile;
