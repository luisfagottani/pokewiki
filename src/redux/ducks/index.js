import { combineReducers } from 'redux';

import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from './app';
import { reducer as PokemonsReducer, REDUCER_NAME as APP_POKEMONS_REDUCER } from './pokemons';
import { reducer as GenerationsReducer, REDUCER_NAME as APP_GENERATIONS_REDUCER } from './generations';

export default combineReducers({
  [APP_REDUCER_NAME]: AppReducer,
  [APP_POKEMONS_REDUCER]: PokemonsReducer,
  [APP_GENERATIONS_REDUCER]: GenerationsReducer,
});
