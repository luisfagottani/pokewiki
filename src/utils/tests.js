import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import { IntlProvider } from 'react-intl';
import { createStore, combineReducers } from 'redux';

import { Messages } from 'i18n';
import middleware, { sagaMiddleware } from 'redux/middleware';

export const createMockStore = ({ reducers, initialState }) => {
  return createStore(combineReducers(reducers), initialState, middleware);
};

export const mockSaga = (fnGenerators) => sagaMiddleware.run(fnGenerators);

function render(ui, store, { ...renderOptions } = {}) {
  const locale = store.getState().app.locale ?? 'EN';
  const history = createMemoryHistory();
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <IntlProvider messages={Messages[locale]} locale={locale}>
          <Router history={history}>{children}</Router>
        </IntlProvider>
      </Provider>
    );
  }

  return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
