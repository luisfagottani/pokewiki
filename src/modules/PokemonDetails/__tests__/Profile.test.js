import { createMockStore, render } from 'utils/tests';
import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from 'redux/ducks/app';
import Profile from '../components/Profile';
import { details } from './__fixtures__';

const mockStore = createMockStore({
  reducers: { [APP_REDUCER_NAME]: AppReducer },
  initialState: {
    [APP_REDUCER_NAME]: {
      isLoading: true,
      locale: 'EN',
    },
  },
});

describe('<PokemonDetails />', () => {
  it('Should render without errors', () => {
    const { container } = render(<Profile name="bulbasaur" details={details} />, mockStore);

    expect(container).toMatchSnapshot();
  });

  it('Should render with the correct info', () => {
    const { getByText } = render(<Profile name="bulbasaur" details={details} />, mockStore);

    getByText('bulbasaur');
    getByText('1');
    getByText('64');
    getByText('7');
    getByText('69');
  });

  describe('if the details are being loaded', () => {
    it('Should show the skeleton loading', () => {
      const { container } = render(<Profile name="bulbasaur" details={undefined} />, mockStore);
      expect(container.querySelector('.list--loading')).toBeTruthy();
    });
  });
});
