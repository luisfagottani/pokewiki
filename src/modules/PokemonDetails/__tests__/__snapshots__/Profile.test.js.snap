// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonDetails /> Should render without errors 1`] = `
<div>
  <div
    class="profile"
    data-testid="profile"
  >
    <figure
      class="profile__figure"
    >
      <figcaption
        class="profile__caption"
      >
        bulbasaur
      </figcaption>
      <img
        alt="bulbasaur"
        class="profile__image"
        src="https://img.pokemondb.net/artwork/bulbasaur.jpg"
      />
    </figure>
    <div
      class="list"
    >
      <dl
        class="list__dl"
      >
        <dt
          class="list__dt"
        >
          National NÂº
        </dt>
        <dd
          class="list__dd"
        >
          1
        </dd>
        <dt
          class="list__dt"
        >
          Base Experience
        </dt>
        <dd
          class="list__dd"
        >
          64
        </dd>
        <dt
          class="list__dt"
        >
          Height
        </dt>
        <dd
          class="list__dd"
        >
          7
        </dd>
        <dt
          class="list__dt"
        >
          Weight
        </dt>
        <dd
          class="list__dd"
        >
          69
        </dd>
      </dl>
    </div>
  </div>
</div>
`;
